pallets/parachain-staking/src/lib.rs

```diff
@@ -438,6 +441,31 @@ pub mod pallet {
 				weight = weight.saturating_add(T::OnNewRound::on_new_round(round.current));
 				// pay all stakers for T::RewardPaymentDelay rounds ago
 				weight = weight.saturating_add(Self::prepare_staking_payouts(round.current));
+
+				// select total candidates
+				let candidates = <CandidatePool<T>>::get().0;
+
+				// iter candidates to check which of them must be marked as offline
+				for candidate in candidates {
+					match <CandidateRoundsInfo<T>>::try_get(candidate.owner.clone()) {
+						Ok(info) => {
+							if round.current - info.0 > T::NumberRoundsOffline::get() {
+								// if the collator has not produced any block within
+								// NumberRoundsOffline e.g(3 rounds for Moonriver)
+								// it is marked as offline
+								<Pallet<T>>::go_offline(T::RuntimeOrigin::from(
+									Some(candidate.owner.clone()).into(),
```

Comment:
An advise: create a function that is called `do_go_offline`, that both the `go_offline` extrinsic and this function call

Reply:
I agree with this remark, usually we not callling runtime calls directly

Reply:
Done!